### ISSUE TITLE
programm output to the console problem.

### ISSUE THREAD

**Issue by wfcloherty** [Alexpux/MSYS2-packages#1565]
the program output of the printf() and puts() funtions do not show in the console until the program exits.

I am using the msys2-x86_64-20180531.exe installer, installed and updated on 1/29/2109. the following packages also installed: base-devel, git, subversion and mingw-w64-x86_64-toolchain. 

The computer is running Windows 10 Pro, also updated to current date. 

Using the following code snippet:
  #include <stdio.h>
  int main(){
    int loop;
    char glyph[3];

    for( loop = 1; loop == 10; loop++){
       printf("Testing printf\n");
    }
    puts("Press any key to exit. ");
    scanf("%2", glyph);
    return 0;
  }

and complied with gcc filename.c -o filename.  (ignore errors in typing if any, I had to retype this into the ticket instead of pasting.) 

when I run the resulting program in the msys2 mingw64 shell the output looks like:

./filename
X                            // I had to press the 'X' key and return at this point. then I get the following.
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Press any key to exit. 

All the printf() and puts() appear to  print after the program exits.   

But if I run it in the native cmd shell of windows I get the correct output of:
filename.exe
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Testing printf
Press any key to exit. X        //this is when I pressed 'X' and return.

All the printf() and puts() actually print while the program is running.   

So is this where I am missing a needed  option to gcc or is something wrong with the environment that the msys2 command shell sets up?

**Comment by StarWolf3000** [comment]
I had a similar problem too, when building and running the tests in vhelin/wla-dx#186 (more specifically, [this comment](https://github.com/vhelin/wla-dx/issues/186#issuecomment-447863766)).

There seem to be cases, when `bash` buffers the output, so you would need to explicitely flush it after each `printf()` and `puts()`, by using `fflush(stdout)`:
```c
#include <stdio.h>

int main() {
  int loop;
  char glyph[3];
  for( loop = 1; loop == 10; loop++){
    printf("Testing printf\n");
    fflush(stdout);
  }
  puts("Press any key to exit. ");
  fflush(stdout);
  scanf("%2", glyph);
  return 0;
}
```

**Comment by wfcloherty** [comment]
StarWolf3000;

Thanks for your reply.  It gave me the lead I needed to solve my issue.

It appears that file streams like STDOUT, STDIN and STDERR and any user defined streams . sent to the console will default to being unbuffered, but when attached to a pipe will changed to being buffered.  This appears to be done by the GLibC library routines that handle streaming.  

Since the default environment for Msys2 uses mintty as its terminal the issue was the STDOUT of my sample program was being piped into mintty, therefore being changed to being buffered. 

My issue was fixed by adding the -defterm argument to the shell.cmd used to start the msys2 environment.
